docs/design/2022-03-31-concurrent-ddl-framawork.md

```diff
@@ -0,0 +1,286 @@
+```
+
+Then, we find some jobs and check Rule 3 and Rule 4:
+
+Rule 3: check if there is a smaller `drop schema` job id on the same schema.
+```sql
+select * from mysql.tidb_ddl_job where is_drop_schema and job_id < {job.id} limit 1;
+```
+Rule 4: if the job is `drop schema`, check if there is a smaller job id on the same schema.
+```sql
+select * from mysql.tidb_ddl_job where job_id < {job.id} limit 1;
+```
+
+If there are records, we can not run the drop schema job.
+
+The workflow will looks like
+
+![concurrent-ddl-framework](./imgs/concurrendl-ddl.png)
+
+compare current implementataion
+
+![queue-base-ddl-framework](./imgs/queue-ddl.png)
+
+### Show DDL jobs
+
+get DDL jobs from the tables:
+```sql
+select * from mysql.tidb_ddl_job;
+select * from mysql.tidb_history_job;
+```
+
+### Cancel DDL
+
+find the cancel job and update the DDL meta to `JobStateCancelling` state
+```sql
+select job_meta from mysql.tidb_ddl_jon where job_id = {job_id};
+-- set the job state to `JobStateCancelling`
+update mysql.tidb_ddl_job set job_meta = {job} where job_id = {job.id}
+```
+
+### Upgrade compatibility
+
+Consider the rolling upgrade, the first upgrade tidb instance will hang if there are internal DDLs because the new version TiDB will write DDL job into table, but the old version TiDB DDL owner will not run the job. So the upgrading TiDB should get the DDL owner first:
+```golang
+//...
+if needDoUpgrade {
+    ddl.requireOwner();
+    // Do upgrade ...
+}
+```
+
+In general, there will have no queueing DDL job when upgrading, but if there are some queueing DDL jobs, TiDB should migrate the DDL jobs from queue to table and wait for DDL job manager handle them.
+
+```golang
+//...
+if needDoUpgrade {
+    ddl.requireOwner();
+    migrateDDL() // migrate queueing DDL jobs from queue to table, including reorg meta.
+    // Do upgrade ...
+}
+```
+
+### Compatibility with CDC
+
+CDC will watch the key range of the ddl job **queue** and sync the finished DDLs to other cluster, after concurrent DDL is implemented, CDC should watch the key range of the `tidb_ddl_job` table and unmarshal the ddl job.
+Since we maintain the old ddl framework at the same time, CDC should be able to watch key range of the ddl job queue and `tidb_ddl_job` table.
+After several versions, CDC can remove the code of watching the key range of the queue.
+
+
+### How to change the tables(`tidb_ddl_job`, `tidb_ddl_reorg`, `tidb_ddl_history`) meta?
+
+If a new field is required in `tidb_ddl_job`, we can use a `ALTER TABLE` statement to add the field, the important thing is make this DDL run successfully.
+Consider the rolling upgrade progress, only the first instance will run the upgrade SQL, at this time, the DDL owner will be the other TiDB instance, so that the DDL will
+be executed successfully by the other old TiDB instance.
+
+## Test Design
+
+We will use [schrddl](https://github.com/PingCAP-QE/schrddl) test the concurrent DDL framework.
```

Comment:
Can you introduce the schrddl ？ How does it test concurrent ddl？

Reply:
is it better to write it in `readme.md` in `schrddl`?

Reply:
We can use a sentence or two to introduce him.

```
Schrddl is a fuzz test tool to generate different DDL or DML jobs to verify TiDB. For more information, you can read this [document](https://github.com/PingCAP-QE/schrddl)
```